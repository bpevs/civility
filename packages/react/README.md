Philosophy
==========
There are only two kinds of UI Elements to familiarize yourself with when using Civility:

**Components**
These never directly consume AppState or ServiceData, and extend from the Civility `Component` class. However, they can access data and methods passed from pages. If a component requires certain data to run, it should throw a warning so that the page can properly subscribe to that data.

**Pages**
These can consume AppState and ServiceData, and extend the Civility `Page` class. This allows us to subscribe to different aspects of AppState.

Think of these class extensions more like prop-injectors; they shouldn't affect the render, just give additional lifecycle and dependency handling.


Usage
=====
```js
import React from "react"
import { Component, Page } from "@civility/react"
import { Comments, Router } from "@civility/store"

// This Example page would just be one comment thread (think Reddit).
class CommentThreadPage extends Page {
  // Inherit methods and data from these stores
  stores = [ Comments, Router ]

  constructor(props) {
    super(props)
  }

  render() {
    const { route, comments } = this.props

    return (
      <div className="comment-thread">
        <Comment root comment={comments[route.threadId]} />
        <Comment.ReplyBox open />
      </div>
    )
  }
}


class Comment extends Component {
  constructor(props, state) {
    super(props, state)
  }

  render() {
    const { content, root, user } = this.props.comment

    return (
      <div className={ root ? "root" : "reply"}>
        <Comment.Author author={user} />
        <Comment.Content content={content} />
        <Comment.Replies replies={replies} />
      </div>
    )
  }
}


Comment.Replies = class CommentReplies extends Component {
  constructor(props, state) {
    super(props, state)
  }

  render() {
    const replies = this.props.replies.map((comment, key) => {
      return <Comment comment={comment} key={key} />
    })

    return <div className="comment-replies">{replies}</div>
  }
}


Comment.ReplyBox = class CommentReplyBox extends Component {
  cache = "session"; // Cache state changes in sessionStorage
  state = {};

  constructor(props, state) {
    super(props, state)
  }

  onChange = evt => this.setState({ text: evt.target.value })

  render() {
    const { open, action } = this.props

    /**
     * When our page subscribes to "comments", it gives Civility components
     * method access to actionsassociated with that Service Data. Some actions,
     * such as leaving comments, require other data, like UserData. The Comment
     * Store will be aware of dependencies such as this. However, unless you
     * specifically subscribe, you won't get explicit data or method access.
     */
    const submit = action.postComment.bind(this)

    return (
      <form className={classNames("comment-reply-box", { open })} submit={submit}>
        <textarea onChange={this.onChange} />
        <button />
      </form>
    )
  }
}
```


Styles
======
Default styles are imported via `@civility/stylesheets`. You can also opt-out of these, but they give a lot of the base styles for components.

The main philosophy for styles is two-fold:
 - A user should be able to theme components via a single imported css stylesheet.
 - A user should always be able to override any default style with 1 custom CSS class.

All UI Elements have a few ways of being styled:
- **You can pass in your own styles, classNames, and ids to any component**. These will be applied to the root component that you are using. This is useful for overriding styles on an individual component instance, and is also useful for creating custom component wrappers.
- **All Components and Pages have built-in classNames** in kebab-case( e.g. `Comment.ReplyBox => comment-reply-box`). There are no styles associated with these classes, and are meant to be used as hooks. This way, it is easy to override an entire class of components at once.
- **All Styled Components and Pages inherit overridable styles from our custom extension of [BassCSS](http://basscss.com/)**. This way, you should be able to override all components to make a theme with a single css file, imported after `@civility/stylesheets`. The classNames created by Civility can look unwieldy, but if you don't want to learn how basscss is used, you don't need to; you can override via the other methods mentioned with a single css class imported after `@civility/stylesheets`.
