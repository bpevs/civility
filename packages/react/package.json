{
  "name": "@civility/react",
  "version": "0.0.1",
  "description": "Civility React Components",
  "main": "./dist/index.js",
  "license": "UNLICENSED",
  "repository": "git+https://github.com/ivebencrazy/civility.git",
  "scripts": {
    "build": "run-p build:* --aggregate-output",
    "build:commonjs": "tsc --outDir './dist/commonjs/' --module 'commonjs'",
    "build:es6": "tsc --outDir './dist/es6/' --module 'es6'",
    "start": "npm run build:es6 -- --watch"
  },
  "dependencies": {
    "lodash-es": "^4.17.7",
    "react": "^16.2.0",
    "react-dom": "^16.2.0"
  },
  "devDependencies": {
    "npm-run-all": "^4.1.2",
    "typescript": "^2.7.2"
  },
  "publishConfig": {
    "access": "public"
  }
}
