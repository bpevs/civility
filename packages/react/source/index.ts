// Components
export * from "./components/Background/Background"
export * from "./components/Button/Button"
export * from "./components/DateTime/DateTime"
export * from "./components/Input/Input"
export * from "./components/Only/Only"
export * from "./components/OnlyJS/OnlyJS"
export * from "./components/NoJS/NoJS"
export * from "./components/SignInForm/SignInForm"

// Hooks
export * from "./hooks/useFocus/useFocus"
export * from "./hooks/useHover/useHover"
