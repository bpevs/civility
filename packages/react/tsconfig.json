{
  "compilerOptions": {
    "allowJs": false,
    "declaration": true,
    "jsx": "react"
  },
  "extends": "../../tsconfig",
  "include": [
    "./source/index.ts"
  ]
}
