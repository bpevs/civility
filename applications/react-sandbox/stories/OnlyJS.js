import * as React from "react"
import { storiesOf } from "@storybook/react"
import { OnlyJS } from "@civility/react"

storiesOf("OnlyJS")
  .add("Render OnlyJS", () => (
    <OnlyJS>Hello</OnlyJS>
  ))
  .add("Render OnlyJS", () => (
    <OnlyJS>Hello</OnlyJS>
  ))
