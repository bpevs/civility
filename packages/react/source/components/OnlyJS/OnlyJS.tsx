import { classNames as cx, isNumber, isString } from "@civility/utilities"
import * as React from "react"


export type OnlyJSProps = React.HTMLProps<HTMLDivElement>


export const OnlyJS: React.FC<OnlyJSProps> = ({
  children,
  ...props
}) => <React.Fragment>{
  React.Children.map(children, child => {
    if (isNumber(child) || isString(child)) {
      return <span className="only-js">{child}</span>
    } else {
       const className = cx("only-js", (child as React.HTMLProps<any>).className)
       return React.cloneElement(child, { ...props, className })
    }
  })
}</React.Fragment>
