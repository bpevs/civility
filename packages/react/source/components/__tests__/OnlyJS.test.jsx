import { shallow } from "enzyme"
import * as React from "react"
import { OnlyJS } from "./OnlyJS"

test("Should Render with only-js className", () => {
  const wrapper = shallow(<OnlyJS><span>hello</span></OnlyJS>)
  expect(wrapper).toMatchSnapshot()
})
