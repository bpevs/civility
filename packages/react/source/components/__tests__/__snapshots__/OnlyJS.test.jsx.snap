// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should Render with only-js className 1`] = `
<Fragment>
  <span
    className="only-js"
    key=".0"
  >
    hello
  </span>
</Fragment>
`;
